src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -223,6 +223,20 @@ internal void ClearAnalyzerExceptionHandlers(DiagnosticAnalyzer analyzer)
             }
         }
 
+        internal bool IsSupportedDiagnostic(DiagnosticAnalyzer analyzer, Diagnostic diagnostic, AnalyzerExecutor analyzerExecutor)
+        {
+            var supportedDescriptors = GetSupportedDiagnosticDescriptors(analyzer, analyzerExecutor);
+            foreach (var descriptor in supportedDescriptors)
```

Comment:
I agree with both of your views.
For compiler analyzer we don't even want to invoke SupportedDiagnostics to determine whether or not to run the analyzer AND whether or not diagnostics it reports are supported. We already did the former, and I will modify the PR to do the later as well by special casing the compiler analyzer.
I also agree with John's comment that for rest of the analyzers, we should be fine doing a linear scan here, doesn't buy us much by adding any more complexity.